test/rnn_seq_api.cpp

```diff
@@ -0,0 +1,176 @@
+            return false;
+
+        if(this->inputMode == 1 && this->hiddenSize != this->inVecLen)
+            return false;
+
+        if((this->rnnMode != 2) && (!this->nocx || !this->nocy))
+            return false;
+
+        if(this->seqLenArray.size() > this->batchSize)
+            return false;
+
+        if(!this->seqLenArray.empty())
+        {
+            if(this->seqLength <
+               *std::max_element(this->seqLenArray.begin(), this->seqLenArray.end()))
+                return false;
+
+            if(this->io_layout == 1)
+            {
+                return std::is_sorted(
+                    this->seqLenArray.begin(), this->seqLenArray.end(), std::greater<int>());
+            }
+        }
+        return true;
+    }
+
+    void run()
+    {
+        if(!this->full_set || (is_correct_params() && !is_skip_comb()))
+            rnn_seq_api_test_driver<T>::run();
+        else
+        {
+            if(this->verbose)
+                std::cout << "Incompatible argument combination, test skipped: "
+                          << this->get_command_args() << std::endl;
+        }
+    }
+};
+
+template <class T>
+struct lstm_MS_solver : rnn_seq_driver<T>
+{
+    lstm_MS_solver() : rnn_seq_driver<T>(true)
+    {
+        std::vector<int> modes(2, 0);
+        modes[1] = 1;
+
+        this->add(this->inVecLen, "vector-len", this->generate_data(std::vector<int>{1, 7}, 7));
+        this->add(this->hiddenSize, "hidden-size", this->generate_data({13, 1}, 13));
+        this->add(this->useDropout, "use-dropout", this->generate_data({0}));
+        this->add(this->numLayers, "num-layers", this->generate_data({3}));
+        this->add(this->inputMode, "in-mode", this->generate_data({0}));
+        this->add(this->biasMode, "bias-mode", this->generate_data(modes));
+        this->add(this->dirMode, "dir-mode", this->generate_data({0}));
+        this->add(this->rnnMode, "rnn-mode", this->generate_data({2}));
+        this->add(this->algoMode, "algo-mode", this->generate_data({0}));
+
+        this->add(this->io_layout, "io_layout", this->generate_data({2}, 2));
+        this->add(this->batchSize, "batch-size", this->generate_data({1, 6}, 6));
+        this->add(this->seqLength, "seq-len", this->generate_data({38}));
+        this->add(this->seqLenArray,
+                  "seqLen-batch",
+                  this->generate_data({
+                      {34, 3, 2, 1},
+                      {1, 15, 34, 15, 34, 1},
+                      {},
+                  }));
+
+        this->add(this->nohx, "nohx", this->generate_data(modes));
+        this->add(this->nocx, "nocx", this->generate_data(modes));
+        this->add(this->nohy, "nohy", this->generate_data(modes));
+        this->add(this->nocy, "nocy", this->generate_data(modes));
+    }
+
+    void run()
+    {
+        if(this->nohx && this->biasMode == 1)
+            return;
+
+        if(this->type == miopenFloat)
+            rnn_seq_driver<T>::run();
```

Comment:
[Q] Is it fine to quietly skip running the test with `--half` etc.?

Reply:
Yes, because this test(lstm_MS_solver) is for a single path in LSTM. And this path is not used in half, because it is not so efficient for it.

Reply:
@shurale-nkn If the test is not intended for half, and this path is not implemented, then maybe the test should explicitly fail when used with half?

Reply:
@atamazov 
Parsing of all arguments, including the data type, takes place inside the test_drive class. 

lstm_MS_solver is a subtest for test [rnn_seq_api.cpp]. if [rnn_seq_api.cpp ] was called it always runs two tests: rnn_seq_driver and lstm_MS_solver. This allows you not to store logic outside of the executable file in cmake. And skipping unused configurations happens relatively quickly and at no additional cost.

Reply:
@atamazov 
Oh, I see what you mean.
Let's just say it won't lead to an error, we can run the test (and rnn API) with these arguments, but then the library will choose another code path (another 'solution') that has already been tested by rnn_seq_driver class . Therefore, in order to save CI time and not to test same code twice, I skip configurations not related to 'lstm_MS_solver' code path.

data_type is the only parameter that is processed not using "struct argument" but directly in test_drive_impl_1, therefore it can only be passed at startup via cmd and cannot be generated as batch_size and others.

And yes, rnn_seq_api have some cheeks for passed parameters, but all these checks are in the general class 'rnn_seq_api_test_driver'. Limitation described in the file 'rnn_seq_api.cpp ' is only used to optimize the size of the test coverage.